services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,191 @@
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for the pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Received invalid message: {}", message);
+            return;
+        } else if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+            throw new MessageException(error);
+        }
+
+        InfoMessage infoMessage = (InfoMessage) message;
+        if (infoMessage instanceof ChunkedInfoMessage) {
+            ChunkedInfoMessage chunked = (ChunkedInfoMessage) infoMessage;
+            processChunkedMessage(chunked);
+        } else {
+            CompletableFuture<InfoMessage> request = pendingRequests.remove(infoMessage.getCorrelationId());
+            if (request != null) {
+                request.complete(infoMessage);
+            } else {
+                logger.warn("Received non-pending message: {}", infoMessage);
+            }
+        }
+    }
+
+    /**
+     * Tries to find and collect all chunked messages into one chain.
+     */
+    private void processChunkedMessage(final ChunkedInfoMessage received) {
+        ChunkedInfoMessage message = received;
+
+        LinkedList<ChunkedInfoMessage> currentChain = messagesChains.stream()
+                .filter(chain -> chain.getLast().getNextRequestId().equals(received.getCorrelationId()))
+                .findAny()
+                .orElse(null);
+
+        while (message != null) {
+            final String correlationId = message.getCorrelationId();
+            ChunkedInfoMessage nextMessage = null;
+
+            if (currentChain != null) {
+                currentChain.add(message);
+            } else {
+                if (pendingChunkedRequests.containsKey(correlationId)) {
+                    logger.trace("Received first message of the chain {}", correlationId);
+                    LinkedList<ChunkedInfoMessage> chain = new LinkedList<>();
+                    chain.add(message);
+                    messagesChains.add(chain);
+
+                    currentChain = chain;
+                } else {
+                    chunkedMessages.put(correlationId, message);
+                }
+            }
+
+            if (currentChain != null) {
+                final String nextCorrelationId = message.getNextRequestId();
+                if (StringUtils.isEmpty(nextCorrelationId)) {
```

Comment:
Comparison with null is more verbose. Also we treat messages with "empty" correlationId as invalid messages... So why should we allow to have an empty string here as a valid value?